### ISSUE TITLE
Rewrite Leo's beautifier in Nim

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#3662]
Rewriting Leo's beautifier in Nim should produce a large speed gain.

Initial experiments with Rust are daunting. PR #3717 is on hold.

This issue depends on #3744: A tokens-only beautifier.

The following Nim program uses [nimpy](https://nimble.directory/pkg/nimpy) to access Python's tokenize module.
This is a strong proof-of-concept for this project.
```nim
import nimpy
let tokenize = pyImport("tokenize")
echo "tokenize module: ", tokenize
```

<details><summary><b>Why I want Orange, not Ruff nor Black</b></summary>
<br>

Black's [magic trailing comma](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#labels-magic-trailing-comma) convention allows fine-grained control over line breaks.

Alas, Black can still breaks lines poorly:

- Black splits definition lines containing trailing comments.
- Black does not allow alignment of key/value pairs in dicts.
- Black splits lines (such as complex conditions) not covered by the magic trailing comma convention.

Orange has other advantages:

- Orange supports `@nobeautify`, allowing columnar formatting of tables.
- Orange does not split/join lines by default.

 **How to work around Black's bad line breaks**

- Use Black to create a baseline format that satisfies **black purists**.
- Fix bad line breaks.
- Use Orange (Leo's beautify command) thereafter.

</details>

**Comment by tbpassin** [comment]
Maybe of interest in this area - 

https://www.blog.pythonlibrary.org/2023/11/27/episode-23-the-ruff-formatter-with-charlie-marsh/

**Comment by edreamleo** [comment]
On Mon, Nov 27, 2023 at 9:39 AM Thomas Passin ***@***.***>
wrote:

> Maybe of interest in this area -
>
>
> https://www.blog.pythonlibrary.org/2023/11/27/episode-23-the-ruff-formatter-with-charlie-marsh/
>
The first comment of #3662
<https://github.com/leo-editor/leo-editor/issues/3662> explains why neither
Black nor Ruff Format will work for me.

Happily, RustPython <https://github.com/RustPython/RustPython>is a solid
foundation for Orange in Rust.

Edward

**Comment by edreamleo** [comment]
I am closing this issue.
